File: code/ch04/4.1.2.CountAndSay.py
class Solution:
    def countAndSay(self, n: int) -> str:
        if n == 1:
            return "1"
        else:
            previous_string = self.countAndSay(n - 1)
            char_index, char_count = 0, 1
            current_string = ""
            for i in range(len(previous_string)):
                if (
                    i + 1 < len(previous_string)
                    and previous_string[char_index] == previous_string[i + 1]
                ):
                    char_count += 1
                else:
                    current_string += str(char_count) + previous_string[char_index]
                    char_index, char_count = i + 1, 1

            return current_string

File: code/ch04/4.1.2.CountAndSay.cpp
class Solution {
public:
    string countAndSay(int n) {
        if (n == 1) return "1";
        
        string pre_str = countAndSay(n - 1);
        
        int char_index = 0, char_count = 1;
        string cur_str = "";
        
        for (int i = 0; i < pre_str.length() - 1; i++) {
            if (pre_str[char_index] == pre_str[i + 1])
                char_count += 1;
            else {
                cur_str += to_string(char_count) + pre_str[char_index];
                char_index = i + 1;
                char_count = 1;
            }
        }
        
        cur_str += to_string(char_count) + pre_str[char_index];
        
        return cur_str;
    }
};
